### PR TITLE
Verify inodes saved on FIM DB before updating file data

### LINKED ISSUE #7829: False positive alerts when multiple files change inode simultaneously
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.0.4+ | Wazuh FIM | Manager/Agent | Packages/Sources | Linux |

## Description
We've identified a condition that leads to false positives being triggered on scans. If we configure a directory to be monitored with `check_mtime="no" check_inode="no"`, then no alert will be triggered if a group of files get deleted and recreated without changes to its contents. Due to how we handle the updates on our local DB, no alert translates to the local DB being outdated with invalid data. This normally is not a big deal, but in the specific case of inodes, it means that the paths are now pointing to a set of metadata which does not correspond to their own, potentially causing a modification of content in one file to trigger succesive alerts on other files.

## Example/How to reproduce
An easier way to see this is with a practical example, lets say we monitor `/mnt/disk1` with the following configuration:
```
<directories check_all="yes" check_mtime="no" check_inode="no">/mnt/disk1</directories>
```

And the contents of the directory are as follows:
```
root@server:/var/ossec/queue/fim/db# ls -lai /mnt/disk1/
total 48
   2 drwxr-xr-x 2 root root 4096 Mar 10 11:02 .
1503 drwxr-xr-x 6 root root 4096 Mar 10 10:03 ..
  11 -rw-r--r-- 1 root root    0 Mar 10 11:02 test
  12 -rw-r--r-- 1 root root  256 Mar 10 11:02 test-0
  13 -rw-r--r-- 1 root root  257 Mar 10 11:02 test-1
  14 -rw-r--r-- 1 root root  258 Mar 10 11:02 test-2
  15 -rw-r--r-- 1 root root  259 Mar 10 11:02 test-3
  16 -rw-r--r-- 1 root root  260 Mar 10 11:02 test-4
  17 -rw-r--r-- 1 root root  261 Mar 10 11:02 test-5
  18 -rw-r--r-- 1 root root  262 Mar 10 11:02 test-6
  19 -rw-r--r-- 1 root root  263 Mar 10 11:02 test-7
  20 -rw-r--r-- 1 root root  264 Mar 10 11:02 test-8
  21 -rw-r--r-- 1 root root  265 Mar 10 11:02 test-9
```

The files have been created with random content. Lets now say that we have a deployment method in place to redeploy the contents of `/mnt/disk1` and the procedure is to delete all of its contents and copy them again from a different place. Once we deploy, contents of `/mnt/disk1` are changed to the following:
```
root@server:/var/ossec/queue/fim/db# ls -lai /mnt/disk1/
total 48
   2 drwxr-xr-x 2 root root 4096 Mar 10 11:05 .
1503 drwxr-xr-x 6 root root 4096 Mar 10 10:03 ..
  11 -rw-r--r-- 1 root root  256 Mar 10 11:05 test-0
  12 -rw-r--r-- 1 root root  257 Mar 10 11:05 test-1
  13 -rw-r--r-- 1 root root  258 Mar 10 11:05 test-2
  14 -rw-r--r-- 1 root root  259 Mar 10 11:05 test-3
  15 -rw-r--r-- 1 root root  260 Mar 10 11:05 test-4
  16 -rw-r--r-- 1 root root  261 Mar 10 11:05 test-5
  17 -rw-r--r-- 1 root root  262 Mar 10 11:05 test-6
  18 -rw-r--r-- 1 root root  263 Mar 10 11:05 test-7
  19 -rw-r--r-- 1 root root  264 Mar 10 11:05 test-8
  20 -rw-r--r-- 1 root root  265 Mar 10 11:05 test-9
```

Effectively, we've only deleted `/mnt/disk1/test`, but due to the fact that all files were deleted and copied again from another directory, they all have their inodes and mtimes modified (which we are not taking into account). After the next scan, `/mnt/disk1/test` will properly be alerted as deleted, but no further changes are made to the DB to adjust the existing files, here are the contents of fim.db at this point:
```
root@server:/var/ossec/queue/fim/db# sqlite3 fim.db "select path, dev, inode, hash_md5 from file_entry inner join file_data on file_entry.inode_id=file_data.rowid order by path"
/mnt/disk1/test-0|2096|12|a378d022def3432674d6fea774c1412a
/mnt/disk1/test-1|2096|13|ad2180330d612f6900b2a6443146cca8
/mnt/disk1/test-2|2096|14|4b3bd1676c4c4885d2184586e427f7de
/mnt/disk1/test-3|2096|15|60e39f5b6a90b5ee121c0bf0a6743450
/mnt/disk1/test-4|2096|16|1583f76f7749da2fa41a481449608fc8
/mnt/disk1/test-5|2096|17|62b27fef8cad16de9e3e9847a87d14af
/mnt/disk1/test-6|2096|18|de54c156c402485a57013eb777787c05
/mnt/disk1/test-7|2096|19|c2b48304c668c0a78488da5a6e45f12b
/mnt/disk1/test-8|2096|20|afc94d4cd821f3a9545e6eff4b08f111
/mnt/disk1/test-9|2096|21|9b8b1eaed0d9a1633bce72a837075c37
```

We can see that the inodes are "off by one", if we modify one of the files, it will cause an alert that is correct, but will store their new metadata in a spot being used by another file which, when scanned, will trigger an alert due to its metadata no longer matching the FS.

I'm attaching the log of the agent in this case, were multiple alerts are shown under this condition
[false-positives.log](https://github.com/wazuh/wazuh/files/6115448/false-positives.log)

## Possible solution
IMO we will need to detach updating the local DB from generating an alert, meaning that even if a given parameter is marked to not be alerted the local DB needs to be updated anyways. Eventually the synchronization mechanism will forward the changes to the manager, also fixing any possible errors that might occur on the inventory.

### PR DESCRIPTION
**Pull request by Molter73**
|Related issue|
|---|
|#7829|

## Description

This PR closes #7829.

A method for preventing files that have had their inodes changed from overwriting metadata from other files has been put in place. This method consists of validating that the dev/inode pair read from the file system is either the same as the file had on the previous scan or the inode obtained is not being used by another file. In case the inode is being used by another file, we scan those files first to find one that has an inode that meets the previous requirements.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
- [x] Source installation
- [x] Source upgrade
- [x] Review logs syntax and correct language
- [x] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
  - [x] ThreadSanitizer

- [x] Added unit tests (for new features)

### CODE REVIEW

**Comment by Molter73** [comment]
Scan-build has found a defect on agent and server compilations, but it is not due to the changes in this PR.
[scan-build-agent.zip](https://github.com/wazuh/wazuh/files/6198000/scan-build-agent.zip)
[scan-build-server.zip](https://github.com/wazuh/wazuh/files/6198001/scan-build-server.zip)

File: src/syscheckd/create_db.c [hunk 1]
```diff
@@ -722,7 +1042,7 @@ int fim_check_depth(const char * path, int dir_position) {
 
 
 // Get data from file
-fim_file_data * fim_get_data(const char *file, fim_element *item) {
+fim_file_data * fim_get_data(const char *file, const fim_element *item) {
     fim_file_data * data = NULL;
```
Comment (jotacarma90) [inline src/syscheckd/create_db.c:1046]:
```suggestion
fim_file_data *fim_get_data(const char *file, const fim_element *item) {
    fim_file_data *data = NULL;
```

**Review by jotacarma90 — CHANGES_REQUESTED** [review]

File: src/syscheckd/db/fim_db_files.c [hunk 2]
```diff
@@ -308,38 +282,61 @@ fim_entry *fim_db_get_path(fdb_t *fim_sql, const char *file_path) {
...
 char **fim_db_get_paths_from_inode(fdb_t *fim_sql, unsigned long int inode, unsigned long int dev) {
+    int i = 0;
     char **paths = NULL;
 
     // Clean statements
     fim_db_clean_stmt(fim_sql, FIMDB_STMT_GET_PATHS_INODE);
-    fim_db_clean_stmt(fim_sql, FIMDB_STMT_GET_PATHS_INODE_COUNT);
+    fim_db_bind_get_inode(fim_sql, FIMDB_STMT_GET_PATHS_INODE, inode, dev);
+
+    os_calloc(2, sizeof(char *), paths);
```
[the remark this answers is not in the document]
Comment (Molter73) [reply src/syscheckd/db/fim_db_files.c:292]: Allocating 1 or 2 pointers is not a major difference in memory usage. Additionally, if the allocated memory is the same as the one being reallocated GNU libc will return the same block, saving us 1 realloc operation.

**Review by jotacarma90 — DISMISSED** [review]

**Comment by Molter73** [comment]
Integration tests on realtime and whodata mode have passed without any failures:
[agent_centos_html_report_test_integration_B1492_20210324124116.zip](https://github.com/wazuh/wazuh/files/6204267/agent_centos_html_report_test_integration_B1492_20210324124116.zip)
[agent_ubuntu_html_report_test_integration_B1492_20210324124116.zip](https://github.com/wazuh/wazuh/files/6204268/agent_ubuntu_html_report_test_integration_B1492_20210324124116.zip)
[agent_windows_html_report_test_integration_B1492_20210324124116.zip](https://github.com/wazuh/wazuh/files/6204269/agent_windows_html_report_test_integration_B1492_20210324124116.zip)

When run on scheduled, Centos and Ubuntu show errors on the synchronization of the FIM DB, but when those same tests are run independently they pass correctly. Looks like a problem in my environment.
[agent_centos_inodes.zip](https://github.com/wazuh/wazuh/files/6204299/agent_centos_inodes.zip)
[agent_ubuntu_inodes.zip](https://github.com/wazuh/wazuh/files/6204305/agent_ubuntu_inodes.zip)

**Review by jotacarma90 — APPROVED** [review]

**Pull request merged**